pkg/query/endpointset.go

```diff
@@ -0,0 +1,704 @@
+// from all endpoints. Keeps around statically defined nodes that were defined with the strict mode.
+func (e *EndpointSet) Update(ctx context.Context) {
+	e.updateMtx.Lock()
+	defer e.updateMtx.Unlock()
+
+	e.endpointsMtx.RLock()
+	endpoints := make(map[string]*endpointRef, len(e.endpoints))
+	for addr, er := range e.endpoints {
+		endpoints[addr] = er
+	}
+	e.endpointsMtx.RUnlock()
+
+	level.Debug(e.logger).Log("msg", "starting to update API endpoints", "cachedEndpoints", len(endpoints))
+
+	activeEndpoints := e.getActiveEndpoints(ctx, endpoints)
+	level.Debug(e.logger).Log("msg", "checked requested endpoints", "activeEndpoints", len(activeEndpoints), "cachedEndpoints", len(endpoints))
+
+	stats := newEndpointAPIStats()
+
+	// Close endpoints which are not active this time (are not in active endpoints map).
+	for addr, er := range endpoints {
+		if _, ok := activeEndpoints[addr]; ok {
+			stats[er.ComponentType()][labelpb.PromLabelSetsToString(er.LabelSets())]++
+			continue
+		}
+
+		er.Close()
+		delete(endpoints, addr)
+		e.updateEndpointStatus(er, errors.New(unhealthyEndpointMessage))
+		level.Info(er.logger).Log("msg", unhealthyEndpointMessage, "address", addr, "extLset", labelpb.PromLabelSetsToString(er.LabelSets()))
+	}
+
+	// Add stores that are not yet in stores.
+	for addr, er := range activeEndpoints {
+		if _, ok := endpoints[addr]; ok {
+			continue
+		}
+
+		extLset := labelpb.PromLabelSetsToString(er.LabelSets())
+
+		// All producers should have unique external labels. While this does not check only StoreAPIs connected to
+		// this querier this allows to notify early user about misconfiguration. Warn only. This is also detectable from metric.
+		if (er.ComponentType() == component.Sidecar || er.ComponentType() == component.Rule) &&
+			stats[component.Sidecar][extLset]+stats[component.Rule][extLset] > 0 {
+
+			level.Warn(e.logger).Log("msg", "found duplicate storeAPI producer (sidecar or ruler). This is not advices as it will malform data in in the same bucket",
+				"address", addr, "extLset", extLset, "duplicates", fmt.Sprintf("%v", stats[component.Sidecar][extLset]+stats[component.Rule][extLset]+1))
+		}
+		stats[er.ComponentType()][extLset]++
+
+		endpoints[addr] = er
+		e.updateEndpointStatus(er, nil)
+
+		if er.HasStoreAPI() {
+			level.Info(e.logger).Log("msg", "adding new storeAPI to query endpointset", "address", addr, "extLset", extLset)
+		}
+
+		if er.HasRulesAPI() {
+			level.Info(e.logger).Log("msg", "adding new rulesAPI to query endpointset", "address", addr)
+		}
+
+		if er.HasExemplarsAPI() {
+			level.Info(e.logger).Log("msg", "adding new exemplarsAPI to query endpointset", "address", addr)
+		}
+
+		if er.HasTargetsAPI() {
+			level.Info(e.logger).Log("msg", "adding new targetsAPI to query endpointset", "address", addr)
+		}
+
+		if er.HasMetricMetadataAPI() {
+			level.Info(e.logger).Log("msg", "adding new MetricMetadataAPI to query endpointset", "address", addr)
```

Comment:
hmhmh, we are not adding anything here ;p just logging I think it might confuse people if some component has all 5 APIs implemented. Users will think we added 5 components but we added one. Maybe more code, but I combine all into one message: `added component with A, B ,C APIs`.

WDYT? 🤗

Reply:
Yes, sound good :)